connection.go

```diff
@@ -873,3 +878,49 @@ func (c *Connection) closeNetwork() {
 func (c *Connection) getLastActivityTime() time.Time {
 	return time.Unix(0, c.lastActivity.Load())
 }
+
+func (c *Connection) sendBufSize() (int, int, error) {
+	if c.sysConn == nil {
+		return 0, 0, fmt.Errorf("no sys call conn")
+	}
+
+	var (
+		// current send buffer usage
+		ioctlErr error
+		sendBuf  = -1
+
+		// current send buffer limit
+		sockoptErr   error
+		sendBufLimit = -1
+	)
+
+	errs := c.sysConn.Control(func(fd uintptr) {
+		sendBuf, ioctlErr = unix.IoctlGetInt(int(fd), unix.SIOCOUTQ)
+		sendBufLimit, sockoptErr = unix.GetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_SNDBUF)
+	})
+
+	errs = multierr.Append(errs, sockoptErr)
+	errs = multierr.Append(errs, ioctlErr)
+	return sendBuf, sendBufLimit, errs
+}
+
+func getSysConn(conn net.Conn, log Logger) syscall.RawConn {
+	type syscallConner interface {
```

Comment:
why define an internal interface instead of type asserting to `tcpConn`?

Reply:
Good question -- I wanted to keep the internal logic in TChannel agnostic to TCP, in case we want to test using unix domain sockets etc in future. Technically, it's possible to pass in a custom `net.Listener` and dialer such that we deal with a `net.Conn` backed by some other transport. Since this logic is no more complex than casting to a `net.TCPConn` while being more generic, thought it was better to keep it generic

Reply:
https://golang.org/pkg/syscall/#Conn seems to implement the same interface?  Should we use this instead?

Reply:
Missed that, updated to use `syscall.Conn`, thanks!